ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/ContractKeysIT.scala

```diff
@@ -0,0 +1,465 @@
+  protected val languageVersion: LanguageVersion = LanguageVersion(LMajV.V1, LMinV.Dev)
+
+  override protected def config: Config = Config.default
+
+  "Working with contract keys" should {
+
+    "permit fetching a divulged contract" in allFixtures { ctx =>
+      def pf(label: String, party: String) =
+        RecordField(label, Some(Value(Value.Sum.Party(party))))
+
+      // TODO currently we run multiple suites with the same sandbox, therefore we must generate
+      // unique keys. This is not so great though, it'd be better to have a clean environment.
+      val key = s"${UUID.randomUUID.toString}-key"
+      val odArgs = Seq(
+        pf("owner", owner),
+        pf("delegate", delegate)
+      )
+      val delegatedCreate = ctx.testingHelpers.simpleCreate(
+        testIdsGenerator.testCommandId("SDVl3"),
+        owner,
+        templateIds.delegated,
+        Record(
+          Some(templateIds.delegated),
+          Seq(pf("owner", owner), RecordField(value = Some(Value(Value.Sum.Text(key))))))
+      )
+      val delegationCreate = ctx.testingHelpers.simpleCreate(
+        testIdsGenerator.testCommandId("SDVl4"),
+        owner,
+        templateIds.delegation,
+        Record(Some(templateIds.delegation), odArgs))
+      val showIdCreate = ctx.testingHelpers.simpleCreate(
+        testIdsGenerator.testCommandId("SDVl5"),
+        owner,
+        templateIds.showDelegated,
+        Record(Some(templateIds.showDelegated), odArgs))
+      for {
+        delegatedEv <- delegatedCreate
+        delegationEv <- delegationCreate
+        showIdEv <- showIdCreate
+        fetchArg = Record(
+          None,
+          Seq(RecordField("", Some(Value(Value.Sum.ContractId(delegatedEv.contractId))))))
+        lookupArg = (expected: Option[String]) =>
+          Record(
+            None,
+            Seq(
+              pf("", owner),
+              RecordField(value = Some(Value(Value.Sum.Text(key)))),
+              RecordField(value = expected match {
+                case None => Value(Value.Sum.Optional(Optional(None)))
+                case Some(cid) => Value(Value.Sum.Optional(Optional(Some(cid.asContractId))))
+              })
+            )
+        )
+        _ <- ctx.testingHelpers.simpleExercise(
+          testIdsGenerator.testCommandId("SDVl6"),
+          submitter = owner,
+          template = templateIds.showDelegated,
+          contractId = showIdEv.contractId,
+          choice = "ShowIt",
+          arg = Value(Value.Sum.Record(fetchArg)),
+        )
+        _ <- ctx.testingHelpers.simpleExercise(
+          testIdsGenerator.testCommandId("SDVl7"),
+          submitter = delegate,
+          template = templateIds.delegation,
+          contractId = delegationEv.contractId,
+          choice = "FetchDelegated",
+          arg = Value(Value.Sum.Record(fetchArg)),
+        )
+        _ <- if (languageVersion precedes LanguageVersion.checkSubmitterInMaintainers) {
```

Comment:
Would it make sense to split the whole test case according to the language features, instead of branching midway through them?

Reply:
It would, but this file is merely extracting the tests related to contract keys from `CommandTransactionChecks.scala`. For that reason I'd rather not modify them at the same time.